pytensor/link/pytorch/dispatch/basic.py

```diff
@@ -56,9 +57,18 @@ def pytorch_funcify_FunctionGraph(
     fgraph_name="pytorch_funcified_fgraph",
     **kwargs,
 ):
+    # Ensure that torch is aware of the leaf nodes
+    # of generated code so we can compile code
+    # without graph breaks
+    def conversion_func_register(*args, **kwargs):
+        functor = pytorch_funcify(*args, **kwargs)
+        module = pytensor.link.utils
+        setattr(module, kwargs["unique_name"](functor), functor)
```

Comment:
Do we need a weakref perhaps? So memory does get freed if nothing else is using these functions?

Reply:
I tried doing weakref (ala `weakref.ref`) in both the leaf node (elemwise dispatch) and in the conversion function (where I current register the module). It seems like link/utils.py has some functionality that takes over the ownership of result - so we would have to make a change to that section. I can give that a shot, but it feels more intrusive (all backends would have weakrefs)

Reply:
The problem with this is PyTensor becomes a giant memory leak if you compile enough pytensor functions?

Can we get input from the torch devs now that we narrowed down the problem?

Reply:
I'll definitely make an ask to the torch team to figure it out, and post here. Probably before we merge this I imagine?

To the point around giant memory leak, I'm not sure that'll be the case to be honest / I'm not particularly worried. These methods work as is if you disable torch compile, so that tells me that the closures are already somewhere. When profiling the script above with memray, I see that we have allocations / memory use within a normal wiggle room (+-5 allocs, 50kb). Here are the zips.
[prof.zip](https://github.com/user-attachments/files/17822043/prof.zip)

Reply:
Isn't this setting am ever increasing number of lambdas as an attribute of that module? 

We once had a problem because we were creating dispatch functions for dynamic Ops in PyMC and someone complained about it (there will always be someone): https://github.com/pymc-devs/pymc/issues/6990

Reply:
Ah, I can explore around with the geographic- but iiuc it just stores another reference to the lambda, it shouldn't be an explicit copy. I could verify by say doing a deep copy and we should see the number go up.

Reply:
But the point is that lambda will never be gc while PyTensor is around no? Even after the torch function disappears?

Reply:
Yup, that's right.

Reply:
I guess it should get gc'd when `link.utils` goes out of scope...unsure when that is

Reply:
This is the issue we end up seeing

```txt
torch._dynamo.exc.InternalTorchDynamoError: weakly-referenced object no longer exists

from user code:
   File "/var/folders/98/g1t2_d2x4w94vqfv06xhyz6c0000gp/T/tmpnk75e974", line 3, in pytorch_funcified_fgraph
    tensor_variable_2, tensor_variable_3 = pytorch_funcified_fgraph(y, z)
```

Reply:
Is there a way to make the functiongraph that torch does keep around to be the one referening those closures? The link.utils will never go away unless someone does `del pytensor`

Reply:
Maybe we can a `__del__`method in the torch linker, and have that cleaned up these explicit references? 

Additionally yea I'll see if I cheese something where we have the global fgraph hold a reference somewhere....

Reply:
I tried to add a sort of wrapper execution function that can just allow us to clean up after ourselves. I do make an assumption that each `pytensor.function` def will have it's own `PytorchLinker`, lmk if that is not correct. This successfully cleaned up the references after the the new function goes out of scope.

Reply:
Yes the function will have it's own PyTorchLinker